tests/test_client.py

```diff
@@ -843,3 +844,78 @@ def test_relation_broken_full(self):
             "standby_gone",
             "database_gone",
         )
+
+
+class TestUpgradeCharm(TestPGSQLBase):
+    def test_leader_migrates_unit_reldata(self):
+        # On upgrade-charm, the leader needs to migrate its unit
+        # relation data to app relation data, which is now preferred.
+        # And mirror it to the shared peer data, where non-leaders
+        # can see it.
+        with self.harness.hooks_disabled():
+            self.harness.set_leader(True)
+            self.harness.update_relation_data(
+                self.relation_id,
+                self.local_unit.name,
+                {"database": "mydb", "extensions": "citext,debversion", "roles": "myrole"},
+            )
+        self.charm.on.upgrade_charm.emit()
+        self.assertEqual(self.relation.data[self.local_app]["database"], "mydb")
+        self.assertEqual(self.relation.data[self.local_app]["extensions"], "citext,debversion")
+        self.assertEqual(self.relation.data[self.local_app]["roles"], "myrole")
+
+    def test_non_leader_waits_to_migrate_unit_reldata(self):
+        # On upgrade-charm, the leader needs to migrate its unit
+        # relation data to app relation data, which is now preferred.
+        # And mirror it to the shared peer data, where non-leaders
+        # can see it.
+        with self.harness.hooks_disabled():
+            self.harness.set_leader(False)
+            self.harness.update_relation_data(
+                self.relation_id,
+                self.local_unit.name,
+                {"database": "mydb", "extensions": "citext,debversion", "roles": "myrole"},
+            )
+
+        self.charm.on.upgrade_charm.emit()
+
+        self.assertNotIn("database", self.relation.data[self.local_app])
+        self.assertNotIn("extensions", self.relation.data[self.local_app])
+        self.assertNotIn("roles", self.relation.data[self.local_app])
+
+        with self.harness.hooks_disabled():
+            self.harness.set_leader(True)
+            # Update the relation data again, to work around
+            # https://github.com/canonical/operator/issues/429
```

Comment:
I see there that leader-elected is triggered according to https://github.com/canonical/operator/issues/429#issuecomment-712308760 however, I failed to see how it can be possible if you do a self.harness.hooks_disabled().

Reply:
Without the hooks_disabled decorator, set_leader(True) emits a leader_elected event. Which was a problem, because it was calling the registered handler and destroying my data (the handler is fine when called as expected, but was unhelpful when called midway through this test setup). Turns out the workaround is not necessary, and the hooks_disabled() context manager is all that is required.

disabling hooks seems like a common thing we will want to do during tests, to ensure that code not part of the tests doesn't get accidentally called mid-test.

Reply:
(the work around has been removed, the hooks_disabled() context manager remains to ensure that only handlers that are part of this test are called)